### ISSUE TITLE
Eager `Any.toString()` in MatcherResult

### ISSUE THREAD

**Issue by PHaroZ** [kotest/kotest#2454]
kotest version 4.6.0

Hi,

Some of my tests deal with huge data class and I found that they are slow & memory consuming because a lots of matchers indirectly call `Any.toString()` even if those tests didn't return any error.

An exemple here : https://github.com/kotest/kotest/blob/0ba14983d485b1b1caef0e5a2cd7402ba47e34d2/kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/nulls/matchers.kt#L104

A simple workarround could be using 
`io.kotest.matchers.MatcherResult.Companion#invoke(boolean, kotlin.jvm.functions.Function0<java.lang.String>, kotlin.jvm.functions.Function0<java.lang.String>)`
instead of 
`io.kotest.matchers.MatcherResult.Companion#invoke(boolean, java.lang.String, java.lang.String)`

The linked example would look like 
```kotlin
    return MatcherResult(passed, { "Expected value to be null, but was $value." }, { "Expected value not be null, but was null." })
```

I could make a PR if you do agree with me.

**Comment by sksamuel** [comment]
Definitely, good spot.

**Comment by sksamuel** [comment]
Fixed in master, will be in 5.0.0.M2
